### PR TITLE
[mono][aot] Enable deduplication of runtime invoke wrappers

### PR DESCRIPTION
**Pull request by kotlarmilos**
This PR contributes to https://github.com/dotnet/runtime/issues/83973. It improves code maintenance by using deduplication enum instead of multiple flags. ~~It enables deduplication of runtime invoke wrappers on iOS and restricts emitting additional methods into the dedup AOT module that are not collected in the `dedup_methods` hash table~~.

~~Here are preliminary results comparing size on disk (SOD) and build time between the baseline (main branch) and the target (this branch).~~

**Update:** `System.Runtime` are failing on the `ios-arm64` CI, even though they pass locally. The failure could be related to missing runtime invoke wrappers during application startup, and the error log may have been masked by a TCP failure. The tests have passed on `tvos-arm64`, which shouldn't differ from `ios-arm64` on the CI. Before merging the deduplication of runtime invoke wrappers on iOS, the CI should be updated to report any error logs that occurred during startup. The table with preliminary size savings on iOS has been moved to the tracking issue.

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -4283,9 +4283,35 @@ get_method_index (MonoAotCompile *acfg, MonoMethod *method)
 	return index - 1;
 }
 
+static gboolean
```
Comment (vargaz) [inline src/mono/mono/mini/aot-compiler.c:4286]: Why are these changes needed ?

Comment (kotlarmilos) [reply src/mono/mono/mini/aot-compiler.c:4286]: The `collect_dedup_method` function encapsulates the logic for collecting deduplication methods, which was previously implemented in the `add_extra_method_full` function. In proposed changes, the logic is moved to `add_method_full` where if a method can be deduplicated, it will return `-1` and the method will not be added to the list of methods which need to be emitted.

Both extra methods and runtime invoke wrappers are added by invoking `add_method_full` to the list of methods which need to be emitted.

Furthermore, when the collected methods need to be emitted in the dedup AOT module, `dedup_emit_mode` ensures that only these methods are emitted, and no other methods are added afterwards. 

Such approach could reduce the size of the emitted code, and ensures that methods that might end up in the dedup AOT module (accidentally) will not be included in the emitted code.

Here is an example:

--- Deduplication Skip Phase ---
When adding methods, `collect_dedup_method` is invoked to check if a method should be skipped.
When adding extra methods, `collect_dedup_method` is invoked to check if a method should be skipped.
When adding wrappers, `collect_dedup_method` is invoked to check if a method should be skipped.

--- Deduplication Include Phase ---
The skipped methods are collected.
dedup_emit_mode is set to TRUE.
When adding collected methods, `collect_dedup_method` allows adding as dedup_emit_mode is set to TRUE.
dedup_emit_mode is set to FALSE.
When adding wrappers (which shouldn't be required), `collect_dedup_method` restricts adding as dedup_emit_mode is set to FALSE.

**Comment by kotlarmilos** [comment]
Relevant failing tests indicate that some wrappers are emitted into the dedup assembly during AOT compilation. These wrappers are later retrieved during runtime by the `mono_aot_get_trampoline` function, which relies on the `mscorlib` AOT module only.

To simplify the build process and improve the SOD, one potential solution is to use `mscorlib` as the dedup assembly. Instead of generating an additional empty dedup assembly, during the collect phase, AOT compiler could collect and emit all deduplicated methods within `mscorlib`. Then, at runtime, `mscorlib` could serve as the container assembly, as it is for wrappers already.

@vargaz @lambdageek @ivanpovazan any thoughts on the proposed approach?

**Comment by lambdageek** [comment]
> Relevant failing tests indicate that some wrappers are emitted into the dedup assembly during AOT compilation. These wrappers are later retrieved during runtime by the `mono_aot_get_trampoline` function, which relies on the `mscorlib` AOT module only.
> 
> To simplify the build process and improve the SOD, one potential solution is to use `mscorlib` as the dedup assembly. Instead of generating an additional empty dedup assembly, during the collect phase, AOT compiler could collect and emit all deduplicated methods within `mscorlib`. Then, at runtime, `mscorlib` could serve as the container assembly, as it is for wrappers already.
> 
> @vargaz @lambdageek @ivanpovazan any thoughts on the proposed approach?

We talked about this briefly during the standup.  I think it's a good idea *in general* to have a uniform approach to where we put "extra methods" during AOT compilation.  So it makes sense to me to have a general mechanism in the AOT compiler for emitting the extra methods that doesn't always just hardcode corelib.   

Separately there's a question of whether corelib could be the destination for all this "extra" stuff when we're in dedup mode.
I think the answer is probably that is not a great idea.

On ios we have to support Mono.framework - which bundles the BCL and its AOT images into something that is shared across apps.  In the future, we might want to support dedup of the final app in this sort of compilation, too.  In that case we couldn't use corelib as the dedup assembly since it's not going to be per-app.

We may want to use the entrypoint assembly as the dedup assembly.  Or continue with the current approach of emitting a separate AOT image with dedup content.

at AOT runtime we could decide based on whether dedup was used whether to look for wrappers in corelib (if there's no dedup, use the existing behavior), or look for them in the dedup asembly (if dedup was turned on - this would be new behavior).

File: src/mono/mono/mini/aot-compiler.c (hunk 1 above)
Comment (lambdageek) [reply src/mono/mono/mini/aot-compiler.c:4286]: I don't like using `dedup_emit_mode` as a global flag to change the state of the compilation in the middle.

I would suggest changing `dedup_emit_mode` to something like a `dedup_emit_phase` enum that could indicate what the state of the compiler is. 

I think there are like three or four values:
```c
enum {
  DEDUP_EMIT_NONE, // dedup is turned off
  DEDUP_EMIT_COLLECT, // dedup is on but this assembly is not the dedup image, so just collect the methods
  DEDUP_EMIT_DEDUP_METHODS, // dedup should emit dedupable_methods
  DEDUP_EMIT_DEDUP_WRAPPERS, // dedup should emit dedupable wrappers but not any more user methods
}
```
(something like that... I'm not totally sure if that's the right phases)

And then make `collect_dedup_method` depend on only on `acfg->dedup_emit_phase` instead of ten different options, so you can call it like this: `collect_dedup_methods (acfg->dedup_emit_phase, method)`.

And add an explicit `dedup_change_phase (MonoAotCompile *acfg, int next_phase) { acg->dedup_emit_phase = next_phase; }` method so that it can be more obvious when you switch from `DEDUP_EMIT_COLLECTED_METHODS` to `DEDUP_EMIT_COLLECTED_WRAPPERS`

**Comment by kotlarmilos** [comment]
Thank you for the detailed explanation.

After further investigation, I've found that during the AOT compilation of deduplicated methods, the `add_referenced_patch` function adds generic instances that are referenced by compiled method. These instances must be emitted as they are skipped in the `dedup-skip` phase.

Therefore, it is not possible to restrict the emitting of additional methods into the dedup AOT module that are not collected in the `dedup_methods` hash table.

With the deduplication of runtime wrappers only, an additional size saving of approximately 1% can be achieved on the HelloiOS sample app. Additionally, introducing a dedup state, as you outlined, can improve code maintenance.

File: src/mono/mono/mini/aot-compiler.c (hunk 1 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/aot-compiler.c:4298]: Thanks for the suggestion, added.

**Review by lambdageek — APPROVED** [review]
LGTM.  @vargaz did you want to look again?

**Comment by kotlarmilos** [comment]
> What is the purpose of the changes to aot-compiler.c ? Is it just refactoring, or it fixes some issue ?

Changes in `aot-compiler.c` refactor the deduplication logic and use enum instead of multiple flags. Additionally, it aims to achieve additional size savings on iOS platforms by deduplicating runtime invoke wrappers.

As a result of deduplicating runtime invoke wrappers on iOS, the `System.Runtime` library tests are passing successfully on a device locally and on `tvos-arm64` on CI. However, the tests are failing only on `ios-arm64` on CI during the startup, so error log isn't available.

First assumption was that `ios-arm64` CI pipeline is using deprecated osx version, but it didn't help. If we didn't manage to reproduce the failure locally, I would revert the changes for the runtime invoke wrappers and merge the PR with enum changes.

Any ideas on why it could fail are beneficial :) Anyway, I will keep you posted here.

File: eng/pipelines/libraries/helix-queues-setup.yml [hunk 2]
```diff
@@ -119,8 +119,6 @@ jobs:
     - ${{ if in(parameters.platform, 'ios_arm64') }}:
       # split traffic for runtime-extra-platforms (which mostly runs on rolling builds)
```
[the remark this answers is not in the document]
Comment (kotlarmilos) [reply eng/pipelines/libraries/helix-queues-setup.yml:120]: I have noticed that the `OSX.1015.Amd64.Iphone.Open` queue is being used for an arm64 device. It is possible that the architecture specified refers to the host instead of the device.

[the remark this answers is not in the document]
Comment (premun) [reply eng/pipelines/libraries/helix-queues-setup.yml:120]: Yes, the `Amd64` refers to the host machine, the `Iphone` part refers to a device that is attached to it and that one is `arm64` (it's a real iPhone device)

**Review by vargaz — APPROVED** [review]

**Comment by kotlarmilos** [comment]
The failures appear to be unrelated. The timeout issue for `tvos-arm64` is being tracked in https://github.com/dotnet/arcade/issues/11683. The failures related to WASM are known https://github.com/dotnet/runtime/issues/84434 and https://github.com/dotnet/runtime/issues/84101.

**Comment by kotlarmilos** [comment]
This PR identified the underlying bug and accidentally introduced a regression, so it was reverted. Currently, there is an open PR https://github.com/dotnet/runtime/pull/85908 that should deduplicate runtime invoke wrappers on iOS. Thanks for the reminder, there is no need for any further actions.
